crates/core/src/client/inner/navigation.rs

```diff
@@ -0,0 +1,485 @@
+            let ext = self.future.drain(entry..);
+            self.history.extend(ext.rev());
+            return Some(id);
+        }
+
+        None
+    }
+
+    /// Returns the current history entry and state
+    pub fn current(&self) -> Option<NavHistoryEntry> {
+        self.history.last().cloned()
+    }
+
+    fn replace_entry(&mut self, history_entry: NavHistoryEntry) {
+        if let Some(last) = self.history.last_mut() {
+            self.id_source += 1;
+
+            *last = history_entry
+        } else {
+            self.push_entry(history_entry)
+        }
+    }
+
+    fn push_entry(&mut self, history_entry: NavHistoryEntry) {
+        self.id_source += 1;
+        self.history.push(history_entry);
+    }
+
+    pub fn set_event_handler(&mut self, handler: Arc<dyn NavEventHandler>) {
+        self.navigation_event_handler.0 = Some(handler)
+    }
+
+    pub fn handle_event(&mut self, event: NavEvent, emit_event: bool) -> HandlerResponse {
+        if !emit_event {
+            return HandlerResponse::Default;
+        }
+
+        if let Some(handler) = self.navigation_event_handler.0.as_ref() {
+            handler.handle_event(event)
+        } else {
+            HandlerResponse::Default
+        }
+    }
+
+    /// create a new destination if one would be added to history, this includes
+    /// the next unique ID that would be issued.
+    fn speculative_next_dest(&self, url: &Url, state: Option<Vec<u8>>) -> NavHistoryEntry {
+        NavHistoryEntry {
+            id: self.id_source + 1,
+            url: url.to_string(),
+            state,
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::sync::Mutex;
+
+    use super::*;
+
+    // Mock event handler used to validate the internal
+    // navigation objects state.
+    pub struct NavigationInspector {
+        last_event: Mutex<Option<NavEvent>>,
+    }
+
+    impl NavEventHandler for NavigationInspector {
+        fn handle_event(&self, event: NavEvent) -> HandlerResponse {
+            *self.last_event.lock().expect("Lock poisoned!") = Some(event);
+            HandlerResponse::Default
+        }
+    }
+
+    impl NavigationInspector {
+        pub fn new() -> Self {
+            Self {
+                last_event: None.into(),
+            }
+        }
+
+        pub fn last_event(&self) -> Option<NavEvent> {
+            self.last_event.lock().expect("Lock poisoned!").clone()
+        }
+    }
+
+    //impl NavEvent {
+    //    // utility function so I can sugar out boiler plate code in tests.
+    //    pub fn empty() -> Self {
+    //        Self {
+    //            to: NavHistoryEntry {
+    //                url: String::new(),
+    //                id: 0,
+    //                state: None,
+    //            },
+    //            event: NavEventType::Push,
+    //            same_document: false,
+    //            from: None,
+    //            info: None,
+    //        }
+    //    }
+    //}
```

Comment:
I think this might be an accident added but looks to be in `crates/core/src/live_socket/tests/navigation.rs` in the working branch of this PR.